helix-term/src/ui/editor.rs

```diff
@@ -698,51 +697,51 @@ impl EditorView {
     /// otherwise.
     fn handle_keymap_event(
         &mut self,
-        mode: Mode,
-        cxt: &mut commands::Context,
-        event: KeyEvent,
-    ) -> Option<KeymapResult> {
-        cxt.editor.autoinfo = None;
-        let key_result = self.keymaps.get_mut(&mode).unwrap().get(event);
-        cxt.editor.autoinfo = key_result.sticky.map(|node| node.infobox());
+        cx: &mut commands::Context,
+        key_result: &KeymapResult,
+    ) -> Option<KeymapResultKind> {
+        cx.editor.autoinfo = key_result.sticky.map(|node| node.infobox());
 
         match &key_result.kind {
-            KeymapResultKind::Matched(command) => command.execute(cxt),
-            KeymapResultKind::Pending(node) => cxt.editor.autoinfo = Some(node.infobox()),
+            KeymapResultKind::Matched(command) => command.execute(cx),
+            KeymapResultKind::Pending(node) => cx.editor.autoinfo = Some(node.infobox()),
             KeymapResultKind::MatchedSequence(commands) => {
                 for command in commands {
-                    command.execute(cxt);
+                    command.execute(cx);
                 }
             }
-            KeymapResultKind::NotFound | KeymapResultKind::Cancelled(_) => return Some(key_result),
+            KeymapResultKind::NotFound => return Some(KeymapResultKind::NotFound),
+            KeymapResultKind::Cancelled(evs) => {
+                return Some(KeymapResultKind::Cancelled(evs.to_vec()))
+            }
         }
         None
     }
 
     fn insert_mode(&mut self, cx: &mut commands::Context, event: KeyEvent) {
-        if let Some(keyresult) = self.handle_keymap_event(Mode::Insert, cx, event) {
-            match keyresult.kind {
-                KeymapResultKind::NotFound => {
-                    if let Some(ch) = event.char() {
-                        commands::insert::insert_char(cx, ch)
-                    }
+        let mut keymaps = self.keymaps().clone();
+        let insert_keys = keymaps.get_mut(&Mode::Insert).unwrap();
```

Comment:
Hmm, this is tricky because some of these commands expect to modify the `keymaps.state`. If you `load()`/`clone()` the entire keymap, then this is applied to a separate instance of the keymap, losing the persistent changes.

You want to only use ArcSwap for the `root` field inside Keymaps: https://github.com/helix-editor/helix/blob/e6a2a605495365b1ac10c1a4f94506de39b70689/helix-term/src/keymap.rs#L331-L334

Reply:
Doing this will also avoid the need to use atomic operations for `pending()` etc., cases where you don't need to interact with the trie

Reply:
I went off the deep end trying to get this to work, and it seems pretty hard at the moment.

It kind of went like:

- Change root to `ArcSwap<KeyTrie>`
  - `PartialEq` fails, change `root` to `KeyTrieRoot(ArcSwap<KeyTrie>)` and implement `PartialEq`
- Had no way to update the `Keymaps` itself anymore
  - Tried out sending an event back up to the event loop for this too
    - Could not re-use the same `ConfigEvent` channel since `Keymaps` isn't in the same crate as `ConfigEvent`, had to make a new channel
    - _I think_ this caused me to start re-writing stuff in a more FP approach internally in Keymaps
      - This introduced bugs when merging keys and I couldn't figure out why

It might be easier to just put `Keymaps` inside an ArcSwap and manually sync it

Reply:
Sorry, I led you down the wrong path.. I thought `Keymap` was the toplevel type, but it turns out to be `Keymaps`.

It seems weird to me that each `Keymap` has a `state` and `sticky` node, I'd move those to the top-level `Keymaps` struct, then you should be able to just swap the hashmap part of `Keymaps`. I can give it a go at solving this?

Reply:
I implemented the changes so it should be easier to make `keymaps.map` a `DynAccess` field :)

Reply:
Sorry actually, I ended up rebasing the commits and pushing them to master. Can you rebase?